### PR TITLE
WIP: Configure fips on the bastion node of openshift cluster

### PR DESCRIPTION
**Pull request by gauravpbankar**
To configure the bastion node of openshfit cluster with FIPS enablement on PowerVM.

- As a part of fips enablement we have already enabled fips on master nodes and worker nodes, In this PR we are going to enable fips on the bastion node.

### CODE REVIEW

File: modules/1_bastion/bastion.tf [hunk 1]
```diff
@@ -345,3 +345,32 @@ resource "null_resource" "setup_nfs_disk" {
     ]
   }
 }
+
```
Comment (prb112) [inline modules/1_bastion/bastion.tf:348]:
```suggestion
```

Duplicate line

File: modules/1_bastion/bastion.tf [hunk 2]
```diff
@@ -345,3 +345,32 @@ resource "null_resource" "setup_nfs_disk" {
...
+
+resource "null_resource" "bastion_nop" {
```
Comment (prb112) [inline modules/1_bastion/bastion.tf]:
```suggestion
resource "null_resource" "fips_enablement" {
```

**Review by prb112 — CHANGES_REQUESTED** [review]
needs changes

File: modules/1_bastion/bastion.tf [hunk 3]
```diff
@@ -345,3 +345,55 @@ resource "null_resource" "setup_nfs_disk" {
...
+    inline = [
+      <<EOF
+# enable FIPS as required
+sudo fips-mode-setup --enable
+EOF
+    ]
+  }
+}
+
+
```
Comment (prb112) [inline modules/1_bastion/bastion.tf]:
```suggestion

```

File: modules/1_bastion/bastion.tf [hunk 4]
```diff
@@ -345,3 +345,55 @@ resource "null_resource" "setup_nfs_disk" {
...
+    private_key = var.private_key
+    agent       = var.ssh_agent
+    timeout     = "${var.connection_timeout}m"
+  }
+
+  provisioner "remote-exec" {
+    inline = [
+      <<EOF
+# rebooting bastion node
+sudo shutdown -r +1
```
Comment (prb112) [inline modules/1_bastion/bastion.tf]:
```suggestion
  count = var.fips_compliant ? local.bastion_count : 0
  depends_on =  [ null_resource.fips_enablement ]
  connection {
    type        = "ssh"
    user        = var.rhel_username
    host        = openstack_compute_instance_v2.bastion[count.index].access_ip_v4
    private_key = var.private_key
    agent       = var.ssh_agent
    timeout     = "${var.connection_timeout}m"
  }
  provisioner "remote-exec" {
    inline = [
      <<EOF
whoami
```

File: modules/1_bastion/bastion.tf [hunk 5]
```diff
@@ -345,3 +345,55 @@ resource "null_resource" "setup_nfs_disk" {
...
+  provisioner "remote-exec" {
+    inline = [
+      <<EOF
+# enable FIPS as required
+sudo fips-mode-setup --enable
+EOF
+    ]
+  }
+}
+
```
Comment (prb112) [inline modules/1_bastion/bastion.tf]:
```suggestion
  count = var.fips_compliant ? local.bastion_count : 0
  depends_on =  [ openstack_compute_keypair_v2.key-pair ,random_id.label ,openstack_compute_flavor_v2.bastion_scg, openstack_compute_instance_v2.bastion, null_resource.bastion_init, null_resource.setup_proxy_info, null_resource.bastion_register, null_resource.enable_repos, openstack_blockstorage_volume_v3.storage_volume, openstack_compute_volume_attach_v2.storage_v_attach, null_resource.bastion_init , null_resource.setup_nfs_disk]
  connection {
    type        = "ssh"
    user        = var.rhel_username
    host        = openstack_compute_instance_v2.bastion[count.index].access_ip_v4
    private_key = var.private_key
    agent       = var.ssh_agent
    timeout     = "${var.connection_timeout}m"
  }
  provisioner "remote-exec" {
    inline = [
      <<EOF
# enable FIPS as required
sudo fips-mode-setup --enable
# rebooting bastion node
sudo shutdown -r +1
EOF
    ]
  }
}

```

File: modules/1_bastion/bastion.tf (hunk 3 above)
Comment (gauravpbankar) [reply modules/1_bastion/bastion.tf]: Added changes.

File: modules/1_bastion/bastion.tf [hunk 6]
```diff
@@ -345,3 +345,53 @@ resource "null_resource" "setup_nfs_disk" {
     ]
   }
 }
+
+resource "null_resource" "fips_enablement" {
+
```
Comment (prb112) [inline modules/1_bastion/bastion.tf]:
```suggestion
```

Comment (prb112) [reply modules/1_bastion/bastion.tf]: target terraform style does not add a return

Comment (prb112) [reply modules/1_bastion/bastion.tf]: please also run a terraform fmt on the code

File: modules/1_bastion/bastion.tf [hunk 7]
```diff
@@ -345,3 +345,53 @@ resource "null_resource" "setup_nfs_disk" {
...
+resource "time_sleep" "wait_60_seconds" {
+  count = var.fips_compliant ? local.bastion_count : 0
+  depends_on = [null_resource.fips_enablement]
+
+  create_duration = "60s"
+}
+
+resource "null_resource" "bastion_nop" {
+
```
Comment (prb112) [inline modules/1_bastion/bastion.tf]: target style does not have a return
```suggestion
```

File: modules/1_bastion/bastion.tf [hunk 8]
```diff
@@ -345,3 +345,53 @@ resource "null_resource" "setup_nfs_disk" {
...
+  }
+}
+
+resource "time_sleep" "wait_60_seconds" {
+  count = var.fips_compliant ? local.bastion_count : 0
+  depends_on = [null_resource.fips_enablement]
+
+  create_duration = "60s"
+}
+
```
Comment (prb112) [inline modules/1_bastion/bastion.tf:377]:
```suggestion

# This step waits for the bastion to come back up and runs a simple command
```

Comment (prb112) [reply modules/1_bastion/bastion.tf:377]: Add comment

File: modules/1_bastion/versions.tf [hunk 9]
```diff
@@ -32,6 +32,9 @@ terraform {
       source  = "hashicorp/random"
       version = "~> 3.4"
     }
+    time = {
+      source = "hashicorp/time"
+      version = "0.9.1"
```
Comment (prb112) [inline modules/1_bastion/versions.tf:37]: This is a new dependency to facilitate synchronization between null_resource providers

File: ocp.tf [hunk 10]
```diff
@@ -83,6 +84,7 @@ module "network" {
 }
 
 module "helpernode" {
+  depends_on = [module.bastion]
```
Comment (prb112) [inline ocp.tf:87]: Added due to module level synchronization problems when FIPS is enabled.

There is an option to use a small dummy step that uses fips_compliant as the synchronization point only when fips is enabled. We opted for this cleaner approach.

**Review by prb112 — CHANGES_REQUESTED** [review]
Please make the few style changes, and change from work in progress to ready for review.

**Review by prb112 — CHANGES_REQUESTED** [review]
comments are inline

**Comment by gauravpbankar** [comment]
@yussufsh : 
Can you please review the PR.

**Comment by yussufsh** [comment]
> @yussufsh : Can you please review the PR.

@gauravpbankar  There are syntax errors in the code. When you are making any further changes I want to see a full run log with the required variables.
Please assign PRs only when you have a working code otherwise it is a waste of time for reviewers.
@prb112

**Comment by prb112** [comment]
Hey @Gaurav / @yussufsh yes - I totally agree. please ping the log to Yussuf and I

**Comment by yussufsh** [comment]
I am going to close this PR, if needed please raise a new one.

**Comment by prb112** [comment]
still needed, new PR will come with new dev

**Pull request closed without merging**